firewood/src/merkle.rs

```diff
@@ -71,11 +83,70 @@ impl<S: ShaleStore<Node>> Merkle<S> {
     }
 }
 
-impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
-    pub fn new(store: Box<S>) -> Self {
-        Self { store }
+impl<'de, S: ShaleStore<Node> + Send + Sync, T: BinarySerde> Merkle<S, T>
+where
+    EncodedNode<T>: serde::Deserialize<'de>,
+{
+    pub fn new(store: Box<S>, serde: T) -> Self {
+        Self { store, serde }
+    }
+
+    // TODO: use `get_encoded` / `decode` instead of `node.encode` / `node.decode` after extention node removal.
+    #[allow(dead_code)]
+    fn get_encoded(&self, node: &ObjRef) -> Result<Vec<u8>, MerkleError> {
+        let encoded = match node.inner() {
+            NodeType::Leaf(n) => EncodedNode {
+                node: EncodedNodeType::Leaf(n.clone()),
+                phantom: PhantomData,
+            },
+            NodeType::Branch(n) => {
+                let mut children: Box<[Option<Vec<u8>>; MAX_CHILDREN]> = Default::default();
+                for (pos, chd) in n.chd().iter().enumerate() {
+                    match chd {
+                        Some(chd_addr) => {
+                            let node = self.get_node(*chd_addr)?;
+
+                            let chd_encoded = self.get_encoded(&node)?;
+                            children[pos] = Some(chd_encoded);
+                        }
+                        None => {
+                            if let Some(v) = &n.children_encoded[pos] {
+                                children[pos] = Some(v.to_vec());
+                            }
+                        }
+                    }
+                }
```

Comment:
In general, I actually find it kind of difficult to read all the branching here. This one isn't that bad, but you have several contexts to keep in your head at any one time. You're on the `Branch` branch of the match statement, then you're looping over children. Now I check to see if the child is avalailable and if not, go look for the pre-encoded child.

What if you used the functional approach instead?

```rust
                // let's look at pairs of children
                let encoded_children = n.chd().iter().zip(n.children_encoded.iter());
                let encoded_children = encoded_children.map(|(child_addr, encoded_child)| {
                    child_addr
                         // if there's a child here, get the encoded bytes
                        .map(|addr| self.get_node(addr).and_then(|node| self.get_encoded(&node)))
                        // or look for the pre-fetched bytes
                        .or_else(|| encoded_child.as_ref().map(|child| Ok(child.to_vec())))
                        .transpose()
                });
                
               // now I don't care about how I got them, but I know I have all my encoded children above

                // now create the empty array
                let mut children: Box<[Option<Vec<u8>>; MAX_CHILDREN]> = Default::default();

                // and just set each child
                for (child, encoded_child) in children.iter_mut().zip(encoded_children) {
                    *child = encoded_child?;
                }
```

With the above code, you deal with any possible branching early on, then just move on. It's just as efficient since iterators are lazy. We don't have to worry about indexing into any array, `zip` takes care of lining everything up for us. 

I know that `map`, `and_then`, `or_else`, and `transpose` aren't the easiest things to understand if you aren't used to these combinators, but it's nice to construct the iterator then think about next steps instead of doing multiple things at once.